File: LC57_InsertInterval/sorting/sorting.py
# Given a set of non-overlapping intervals, insert a new interval into the intervals (merge if necessary).
# You may assume that the intervals were initially sorted according to their start times.
# Input: intervals = [[1,3],[6,9]], newInterval = [2,5]
# Output: [[1,5],[6,9]]
# Input: intervals = [[1,2],[3,5],[6,7],[8,10],[12,16]], newInterval = [4,8]
# Output: [[1,2],[3,10],[12,16]]
# Explanation: Because the new interval [4,8] overlaps with [3,5],[6,7],[8,10].
# Input: intervals = [], newInterval = [5,7]
# Output: [[5,7]]
# Input: intervals = [[1,5]], newInterval = [2,3]
# Output: [[1,5]]
# Input: intervals = [[1,5]], newInterval = [2,7]
# Output: [[1,7]]

from typing import List


def insert(intervals: List[List[int]], newInterval: List[int]) -> List[List[int]]:
    # Sorting
    # Time Complexity: O(nlogn)
    # Space Complexity: O(n)
    intervals.append(newInterval)

    intervals.sort(key = lambda x: x[0])

    result = []

    for interval in intervals:
        if len(result) == 0:
            result.append(interval)
        elif result[-1][1] < interval[0]:
            result.append(interval)
        else:
            result[-1][1] = max(result[-1][1], interval[1])
    
    return result



# # # # # # # # # # # # # # # # # # # # Test Cases # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # 
print("test 1: ")
print(insert([[1,3],[6,9]], [2,5]))

print("test 2: ")
print(insert([[1,2],[3,5],[6,7],[8,10],[12,16]], [4,8]))

print("test 3: ")
print(insert([], [5,7]))

print("test 4: ")
print(insert([[1,5]], [2,3]))

print("test 5: ")
print(insert([[1,5]], [2,7]))

File: LC56_MergeIntervals/Sort/sorting.cpp
/*
* Given an array of intervals where intervals[i] = [starti, endi], merge all overlapping intervals, and return an array of the non-overlapping intervals that cover all the intervals in the input.
* Input: intervals = [[1,3],[2,6],[8,10],[15,18]]
* Output: [[1,6],[8,10],[15,18]]
* Explanation: Since intervals [1,3] and [2,6] overlaps, merge them into [1,6].
* Input: intervals = [[1,4],[4,5]]
* Output: [[1,5]]
* Explanation: Intervals [1,4] and [4,5] are considered overlapping.
*/

#include <iostream>
#include <stdlib.h>
#include <vector>
#include <algorithm>
using namespace std;

bool cmp(vector<int> a, vector<int> b){
    return a[0] < b[0];
}

vector<vector<int>> merge(vector<vector<int>>& intervals) {
    // Sort
    // Time Complexity: O(nlogn)
    // Space Complexity: O(1)
    sort(intervals.begin(), intervals.end(), cmp);

    vector<vector<int>> result;

    for(auto interval : intervals){
        if(result.empty()){
            result.push_back(interval);
        }else if(result.back()[1] >= interval[0]){
            result.back()[1] = max(result.back()[1], interval[1]);
        }else{
            result.push_back(interval);
        }
    }

    return result;
}

int main(){
    // test 1
    cout << "test 1: " << endl;
    vector<vector<int>> test1{{1,3},{2,6},{8,10},{15,18}};
    vector<vector<int>> result = merge(test1); // Expect [[1, 6], [8, 10], [15, 18]]
    for(int i = 0; i<result.size(); i++){
        cout << "[";
        for(int j=0; j < result[0].size(); j++){
            cout << result[i][j] << ", ";
        }
        cout << "]" << endl;
    }

    // test 2
    cout << "test 2: " << endl;
    vector<vector<int>> test2{{15,18},{8,10},{2,6},{1,3}};
    result = merge(test2); // Expect [[1, 6], [8, 10], [15, 18]]
    for(int i = 0; i<result.size(); i++){
        cout << "[";
        for(int j=0; j < result[0].size(); j++){
            cout << result[i][j] << ", ";
        }
        cout << "]" << endl;
    }
    
    // test 3
    cout << "test 3: " << endl;
    vector<vector<int>> test3{{1,4},{4,5}};
    result = merge(test3); // Expect [[1, 5]]
    for(int i = 0; i<result.size(); i++){
        cout << "[";
        for(int j=0; j < result[0].size(); j++){
            cout << result[i][j] << ", ";
        }
        cout << "]" << endl;
    }
    return 0;
}